hax/treemakers/lones2info.py

```diff
@@ -0,0 +1,228 @@
+"""
+Treemakers used in the s2-only analysis
+see https://github.com/XENON1T/s2only for more information
+"""
+
+import hax
+import numpy as np
+
+try:
+    from s2only.common import cathode_z_range
+except ImportError:
+    # Don't have the s2only framework itself
+    # Let's hope this value doesn't change or we remember to change it
+    cathode_z_range = np.array([-100, -95])
+
+
+def get_small_events(data):
+    """Return rows from data corresponding to events
+    that have S2 < 500 and S1 < 70 (or lack S1s altogether).
+    """
+    d = data[data['s2_area'] < 500]
+
+    # Remove a very large population of events with large S1s
+    # where the S2 is probably just fake nonsense after the S1 (AP, PI SE)
+    d = d[
+        np.isnan(d['s1_area'])
+        | (d['s1_area'] < 70)]
+
+    return d
+
+
+class BigBeforeSmall(hax.minitrees.MultipleRowExtractor):
+    """Return very basic info about all large S2s closely preceding a small-S2 event
+    "Large" means: S2 area > max(300 PE, 0.1 * largest_s2_area)
+    "Small" is defined by get_small_events
+    """
+    __version__ = '0.0.7'
+    extra_branches = ['peaks.reconstructed_positions*',
+                      'peaks.left', 'peaks.center_time']
+
+    def get_data(self, dataset, event_list=None):
+        if event_list is not None:
+            raise RuntimeError("BigBeforeSmall computes its own event list")
+
+        # Notice we use bypass_blinding here: we want *all* large events
+        # before small ones, even those that are candidates for funny processes
+        # like DEC and DBD.
+        # (looking back at this, this is probably not necessary, if I recall correctly;
+        # blinding only gets activated if you load the correction minitree
+        # which we do not)
+        data = hax.minitrees.load_single_dataset(
+            dataset,
+            ['Fundamentals', 'LargestPeakProperties'],
+            bypass_blinding=True)[0]
+
+        small_events = get_small_events(data)
+
+        # Find all events for which a small event follows in 1s
+        # This is about 10% of all events in a normal background dataset
+        dts = (small_events['event_time'].values.reshape(1, -1) -
+               data['event_time'].values.reshape(-1, 1))
+        close_mask = np.any((0 < dts) & (dts < int(1e9)), axis=1)
+        event_list = np.where(close_mask)[0]
+
+        return hax.minitrees.MultipleRowExtractor.get_data(self, dataset, event_list=event_list)
+
+    def extract_data(self, event):
+        # Get the largest S2s (up to five) satisfying
+        #   area > max(300 PE, 0.1 * largest_s2_area)
+        s2s = [p
+               for p in event.peaks
+               if p.type == 's2']
+        if not len(s2s):
+            # Hm, that's somewhat interesting...
+            # Maybe have a look at these sometime.
+            return []
+
+        s2_areas = np.array([p.area for p in s2s])
```

Comment:
Can we add x, y corrected S2 for top and bottom array so that we can use cS2Asymmetry for cuts? I guess it's done in analysis notebook if not here, then it's fine. But putting it here might be nicer to reduce complexity in analysis scripts.

Reply:
Yes, we use the (x,y)-corrected AFT for the AFT cut, which we indeed compute in the framework. The computation is done automatically whenever the AFT cut is activated (see https://github.com/XENON1T/s2only/blob/master/s2only/utils.py#L129).

Reply:
ok, it's fine but another analyzer will need to implement it him/herself